addons/partner_autocomplete/models/res_company.py

```diff
@@ -1,14 +1,128 @@
 # -*- coding: utf-8 -*-
 # Part of Odoo. See LICENSE file for full copyright and licensing details.
 
-from odoo import api, fields, models
+from ast import literal_eval
+from urllib.parse import urlparse
+
+import json
+import base64
+import logging
+import requests
+
+from odoo.addons.iap.tools import iap_tools
+from odoo import fields, models, tools, _
+
+_logger = logging.getLogger(__name__)
+
+COMPANY_AC_TIMEOUT = 5
+
 
 class ResCompany(models.Model):
     _name = 'res.company'
     _inherit = 'res.company'
 
     partner_gid = fields.Integer('Company database ID', related="partner_id.partner_gid", inverse="_inverse_partner_gid", store=True)
+    iap_enrich_auto_done = fields.Boolean('Enrich Done')
 
     def _inverse_partner_gid(self):
         for company in self:
             company.partner_id.partner_gid = company.partner_gid
+
+    def iap_enrich_auto(self):
+        """ Enrich company. This method should be called by automatic processes
+        and a protection is added to avoid doing enrich in a loop. """
+        if self.env.user.has_group('base.group_system'):
+            for company in self.filtered(lambda company: not company.iap_enrich_auto_done):
+                company._enrich()
+            self.iap_enrich_auto_done = True
+        return True
+
+    def _enrich(self):
+        """
+        This method calls the partner autocomplete service from IAP to enrich the partner related fields
+        of the company.
+        """
+        self.ensure_one()
+        _logger.info("Starting enrich of company %s (%s)", self.name, self.id)
+
+        company_domain = self._get_company_domain()
+        if not company_domain:
+            return {'error': _("Could not extract company domain. Impossible to enrich the company.")}
```

Comment:
If we're going to log the start of the process, woulnd't logging the issues make sense? Also specifying *which* company here could be a good idea depending on caller.

Reply:
Also why is this returning an apparently never used error message? And never anything useful?

Reply:
I think duplicate of existing IAP code.